proto/p4/p4types.proto

```diff
@@ -0,0 +1,207 @@
+// refer to the types by name in the P4DataTypeSpec message. We also support
+// annotations for these type specs which can be useful, e.g. to identify
+// well-known headers (such as ipv4).
+message P4TypeInfo {
+  map<string, P4StructTypeSpec> structs = 1;
+  map<string, P4HeaderTypeSpec> headers = 2;
+  map<string, P4HeaderUnionTypeSpec> header_unions = 3;
+  map<string, P4EnumTypeSpec> enums = 4;
+  P4ErrorTypeSpec error = 5;
+}
+
+// Describes a P4_16 type.
+message P4DataTypeSpec {
+  oneof type_spec {
+    P4BitstringLikeTypeSpec bitstring = 1;
+    P4BoolType bool = 2;
+    // From the P4_16 spec: "A tuple is similar to a struct, in that it holds
+    // multiple values. Unlike a struct type, tuples have no named fields."
+    P4TupleTypeSpec tuple = 3;
+    P4NamedType struct = 4;
+    P4NamedType header = 5;
+    P4NamedType header_union = 6;
+    P4HeaderStackTypeSpec header_stack = 7;
+    P4HeaderUnionStackTypeSpec header_union_stack = 8;
+    P4NamedType enum = 9;
+    P4ErrorType error = 10;
+  }
+}
+
+message P4NamedType {
+  string name = 1;
+}
+
+// Empty message as no type information needed, just used as a placeholder in
+// the oneof to identify boolean types.
+message P4BoolType { }
+
+message P4ErrorType { }
+
+message P4BitstringLikeTypeSpec {
+  oneof type_spec {
+    P4BitTypeSpec bit = 1;  // bit<W>
+    P4IntTypeSpec int = 2;  // int<W>
+    P4VarbitTypeSpec varbit = 3;  // varbit<W>
+  }
+  // Useful to identify well-know types, such as IP address or Ethernet MAC
+  // address.
+  repeated string annotations = 4;
+}
+
+message P4BitTypeSpec {
+  int32 bitwidth = 1;
+}
+
+message P4IntTypeSpec {
+  int32 bitwidth = 1;
+}
+
+message P4VarbitTypeSpec {
+  int32 max_bitwidth = 1;
+}
+
+message P4TupleTypeSpec {
+  repeated P4DataTypeSpec members = 1;
+}
+
+message P4StructTypeSpec {
+  message Member {
+    string name = 1;
+    P4DataTypeSpec type_spec = 2;
+  }
+  repeated Member members = 1;
+  repeated string annotations = 2;
+}
+
+message P4HeaderTypeSpec {
+  message Member {
+    string name = 1;
+    P4BitstringLikeTypeSpec type_spec = 2;
+  }
+  repeated Member members = 1;
+  repeated string annotations = 2;
+}
+
+message P4HeaderUnionTypeSpec {
+  message Member {
+    string name = 1;
+    P4NamedType header = 2;
+  }
+  repeated Member members = 1;
+  repeated string annotations = 2;
+}
+
+message P4HeaderStackTypeSpec {
+  P4NamedType header = 1;
+  int32 size = 2;
+}
+
+message P4HeaderUnionStackTypeSpec {
+  P4NamedType header_union = 1;
+  int32 size = 2;
+}
+
+// As of today, P4_16 enum members have no integer value assigned to
+// them. Future versions of this protobuf interface may use integer values for
+// enums in P4Data, which means the repeated field would be replaced by a map in
+// this message (mapping from member names to integer values). This is dependent
+// on future versions of the P4 language.
+message P4EnumTypeSpec {
+  repeated string members = 1;
```

Comment:
What about the ability to annotate individual members of the enum?  Does the language allow it?

Reply:
My reasoning was that I would not try to anticipate where annotations could be useful. So I only put annotations where I knew for sure they could be useful. Maybe we can plan on adding the annotations field where needed when the need arises? This can be done easily in a backward-compatible way.

Reply:
Per the P4_16 spec section 7.2.1, enums can be annotated:
```
enumDeclaration
    : optAnnotations ENUM name '{' identifierList '}'
```
At some point I think carrying these annotations into P4Info would be valuable.  I have several ideas related to automatic user-interface generation where annotating an enum (and specifying a vaue for  it)  could be nice.

Reply:
Thanks for the feedback. My proposal already supports annotations on enums, just not on enum *members* (not that the language syntax does not support these either). I wanted to wait until there was interest before adding support for annotations on enum members. However, since both you and Tom have brought-up this point, I have updated my proposal to support annotations on enum members. It will indeed be useful to assign integer values to enum members until the LGDW standardizes something.

Reply:
@antoninbas It looks like https://github.com/p4lang/PI/commit/7cb4fb5de63a4c39c4e1dfc6b7fd4e92a81f4dbe reverted your change?

Reply:
You're right, it looks OK in the final PR. I got lost in the rabbit hole. Thanks!

Reply:
@antoninbas I can't seem to confirm that P4_16 accepts annotations on enum _members_ or more generally identifiers in an identifier list.
```
enumDeclaration
    : optAnnotations ENUM name '{' identifierList '}'
...
identifierList
: name
| identifierList ',' name
;
```